File: 2020/7.rb
lines = File.open("7.txt").read.split("\n")#.map(&:to_i)
rules = Hash.new { |hash, key| hash[key] = [] }
contains = Hash.new { |hash, key| hash[key] = [] }
lines.each do |line|
	bags = line.split(/bags?/)
	bags.map! {|x| x.strip}.select! {|x| x.size > 1}
	bags[1..bags.size].each do |bag|
		count, color = bag.match(/(\d*)? (\w+ \w+)$/).captures
		rules[color] << bags[0]
		contains[bags[0]] << [count.to_i, color] if color != "no other"
	end
end

def count(frontier, rules, contains, p2=false)
	seen = []
	total = 0
	while frontier.size != 0
		temp = frontier.pop
		bagCount, bag = !p2 ? [0, temp]: temp
		if p2 || !seen.include?(bag)
			temp = !p2 ? rules[bag]: contains[bag]
			total += bagCount
			temp.each do |x|
				frontier << (!p2 ? x: [x[0] * bagCount, x[1]])
			end
			seen << bag
		end
	end
	return !p2 ? seen.size: total
end
p count(rules["shiny gold"].clone, rules, contains)
p count(contains["shiny gold"].clone, rules, contains, true)

File: 2020/7.py
import re

lines = open('7.txt').read().split('\n')

def p2(bag, count, multiplier, rules):
    if len(rules[bag]) == 0: return count
    bags_in_this_bag = 0
    for sub_bag in rules[bag]:
        num_sub_bags = multiplier * int(sub_bag[0])
        bags_in_this_bag += p2(sub_bag[1], num_sub_bags, num_sub_bags, rules)
    return bags_in_this_bag + count 

rules = {}
eventually_contain = {}

for line in lines:
    split_line = line.split(" bags contain ")
    bag, sub_bags = split_line[0], ''.join(split_line[1:])
    rules[bag] = []
    for sub_bag in re.findall(r'(\d+) (\w+ \w+) bag', sub_bags):
        rules[bag].append(sub_bag)


# I could break when I see a shiny gold in the while, but I thought it would be more useful to
# create a full list of every bag
for bag, sub_bags in rules.items():
    frontier = [x for x in sub_bags]
    seen = set()
    while len(frontier) > 0:
        popped_bag = frontier.pop()[1]
        if popped_bag not in seen:
            seen.add(popped_bag)
            frontier += rules[popped_bag]
    eventually_contain[bag] = seen

# part 1: using default dict messed me up a little bit causing some hard to find bugs
print (sum(1 for x in eventually_contain.values() if "shiny gold" in x))
# part 2: Forgot to add current number of bags to number of bags in current bag
# (i understand thats confusing)
print (p2("shiny gold", 0, 1, rules))